src/NuGet.Clients/PackageManagement.VisualStudio/IDE/VsProjectReferenceUtility.cs

```diff
@@ -0,0 +1,437 @@
+        {
+            // DTE calls need to be done from the main thread
+            await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+
+            var results = new HashSet<ExternalProjectReference>();
+
+            // projects to walk - DFS
+            var toProcess = new Stack<DTEReference>();
+
+            // keep track of found projects to avoid duplicates
+            var rootProjectPath = EnvDTEProjectUtility.GetFullProjectPath(rootProject);
+
+            // start with the current project
+            toProcess.Push(new DTEReference(rootProject, rootProjectPath));
+
+            var itemsFactory = ServiceLocator.GetInstance<IVsEnumHierarchyItemsFactory>();
+            var uniqueNames = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+
+            // continue walking all project references until we run out
+            while (toProcess.Count > 0)
+            {
+                var dteReference = toProcess.Pop();
+
+                // Find the path of the current project
+                var projectPath = dteReference.Path;
+                var project = dteReference.Project;
+
+                if (string.IsNullOrEmpty(projectPath) || !uniqueNames.Add(projectPath))
+                {
+                    // This has already been processed or does not exist
+                    continue;
+                }
+
+                IReadOnlyList<ExternalProjectReference> cacheReferences;
+                if (context.ClosureCache.TryGetValue(projectPath, out cacheReferences))
+                {
+                    // The cached value contains the entire closure, add it to the results and skip
+                    // all child references.
+                    results.UnionWith(cacheReferences);
+                }
+                else
+                {
+                    // Get direct references
+                    var projectResult = GetDirectProjectReferences(
+                        dteReference.Project,
+                        projectPath,
+                        itemsFactory,
+                        rootProjectPath,
+                        context.Logger);
+
+                    // Add results to the closure
+                    results.UnionWith(projectResult.Processed);
+
+                    // Continue processing
+                    foreach (var item in projectResult.ToProcess)
+                    {
+                        toProcess.Push(item);
+                    }
+                }
+            }
+
+            // Cache the results for this project and every child project which has not been cached
+            foreach (var project in results)
+            {
+                if (!context.ClosureCache.ContainsKey(project.MSBuildProjectPath))
+                {
+                    var closure = DependencyGraphProjectCacheUtility.GetExternalClosure(project.UniqueName, results);
+                    var direct = DependencyGraphProjectCacheUtility.GetDirectReferences(project.UniqueName, closure);
```

Comment:
Spoke offline about this. The existing implementation is depth-first search, meaning that all of the process direct references are not available in any single scope. We decided it would be better to keep existing functionality (DFS) and optimized the `GetDirectReferences` to uses a dictionary, given the method linear time with respect to the number of children.